### PR TITLE
`DT[,foo:=bar]`: earlier check for `selfrefok`

### PR DESCRIPTION
**Pull request by aitap**
Kludge towards #7498; amends #6789 and an earlier half-fix, #7488. Maybe it gives someone a better idea, I don't know. It does fix `bbotk`.

The check needs to happen before `eval(jsub)` / `.Call(dogroups)` because the latter will try to assign the results into the table, potentially trying to resize the number of columns.

Use static analysis to detect common forms of column deletion: `foo := NULL` and `c("bar", "baz") := .(NULL, NULL)`. Static analysis is doomed to miss things like `frob := if (runif(1) < .5) 42 else NULL`, but hopefully it covers the needs of our reverse dependencies.

### CODE REVIEW

**Comment by aitap** [comment]
Fixes: `bbotk`, `blocking`, `circhelp`, `hicp`, `metR`, `mlr3tuning`, `SeaVal`, `slim`, `tidyrules`, `tidytable`, `ume`, `webtrackR`, `DramaAnalysis`.

<details>

Before:

```r
dir.create('extrastuff/Library', FALSE, TRUE)
p <- c("bbotk", "blocking", "circhelp", "hicp", "metR", "mlr3tuning",
"SeaVal", "slim", "tidyrules", "tidytable", "ume", "webtrackR",
"DramaAnalysis")
download.packages(p, 'extrastuff')
install.packages('https://cran.r-project.org/incoming/waiting/data.table_1.18.0.tar.gz', lib = 'extras
tuff/Library')
tools::check_packages_in_dir('extrastuff', xvfb=TRUE, Ncpus=13)
tools::check_packages_in_dir_details('extrastuff') -> de_1.18
de_1.18 |> within(Status <- ordered(Status, c('OK','NOTE','WARNING','ERROR'))) |> aggregate(Status ~ Package, max)
```
```
         Package Status
1          bbotk  ERROR
2       blocking  ERROR
3       circhelp  ERROR
4  DramaAnalysis  ERROR
5           hicp  ERROR
6           metR  ERROR
7     mlr3tuning  ERROR
8         SeaVal  ERROR
9           slim  ERROR
10     tidyrules  ERROR
11     tidytable  ERROR
12           ume  ERROR
13     webtrackR  ERROR
```

After:

```r
install.packages('data.table_1.17.99.tar.gz', lib = 'extrastuff/Library')
tools::check_packages_in_dir('extrastuff', xvfb=TRUE, Ncpus=13)
tools::check_packages_in_dir_details('extrastuff') -> de_pr7502
de_pr7502 |> within(Status <- ordered(Status, c('OK','NOTE','WARNING','ERROR'))) |> aggregate(Status ~ Package, max)
```
```
         Package  Status
1          bbotk      OK
2       blocking      OK
3       circhelp    NOTE
4  DramaAnalysis      OK
5           hicp      OK
6           metR      OK
7     mlr3tuning      OK
8         SeaVal      OK
9           slim WARNING
10     tidyrules      OK
11     tidytable      OK
12           ume      OK
13     webtrackR      OK
```
```r
subset(de_pr7502, Status != 'OK') # unrelated to data.table
```

```
Package: circhelp 1.1
Check: DESCRIPTION meta-information, Result: NOTE
    Missing dependency on R >= 4.1.0 because package code uses the pipe
    |> or function shorthand \(...) syntax added in R 4.1.0.
    File(s) using such syntax:
      ‘functions.R’

Package: slim 0.1.1
Check: Rd files, Result: NOTE
  checkRd: (-1) fit_slim.Rd:24: Lost braces; missing escapes or markup?
      24 | sum(x' V^{-1} (y - x beta)) = 0 as the covariance V tends from V[, , 1] +
         |          ^

Package: slim 0.1.1
Check: re-building of vignette outputs, Result: WARNING
  Error(s) in re-building vignettes:
    ...
  --- пересоздаю ‘slim.Rmd’ с использованием rmarkdown
  ! LaTeX Error: Unicode character З (U+0417)
                 not set up for use with LaTeX.

  Try other LaTeX engines instead (e.g., xelatex) if you are using pdflatex. See https://bookdown.org/yihui/rmarkdown-cookbook/latex-unicode.html
  Ошибка: выполнение очерка 'slim.Rmd' не удалось с диагностикой:
  LaTeX failed to compile /home/ivan/vcs/data.table/extrastuff/slim.Rcheck/vign_test/slim/vignettes/slim.tex. See https://yihui.org/tinytex/r/#debugging for debugging tips. See slim.log for more info.
  --- не удалось пересоздать ‘slim.Rmd’

  РЕЗУЛЬТАТ: обработка следующих файлов не удалась:
    ‘slim.Rmd’

  Ошибка: Vignette re-building failed.
  Выполнение остановлено
```

</details>

What if we checked for new columns before ordinary/by-group evaluation and additionally checked for column removal after ordinary evaluation? `x[, foo := NULL, by = bar]` is not something people normally run, right?

File: R/data.table.R [hunk 1]
```diff
@@ -1220,11 +1220,51 @@ replace_dot_alias = function(e) {
...
             if (is.data.table(x)) warningf("A shallow copy of this data.table was taken so that := can add or remove %d columns by reference. At an earlier point, this data.table was copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. It's also not unusual for data.table-agnostic packages to produce tables affected by this issue. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.", length(newnames))
         }
+        # ok <- selfrefok above called without verbose -- only activated when
+        #   ok=-1 which will trigger setalloccol with verbose in the next
+        #   branch, which again calls _selfrefok and returns the message then
+        # !is.data.table for DF |> DT(,:=) tests 2212.16-19 (#5113) where a shallow copy is routine for data.frame
+        if (
+          (
+            !is.null(newnames) || # adding new columns
```
Comment (ben-schwen) [inline R/data.table.R:1234]: IIRC there is also path with `set` where we add columns where `newnames` is `NULL`

Comment (aitap) [reply R/data.table.R:1234]: Sorry, can't figure it out. Any pointers?

Comment (HughParsonage) [reply R/data.table.R:1234]: I think we just need to do the same thing in `set` (lines 2849-2853) -- which is being done in #7500

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -21909,3 +21909,8 @@ DT = suppressMessages(rbindlist(DTl)) # used to crash
 test(2352.1, dim(DT), c(300002L, 1L))
 test(2352.2, DT[[1]], rep(42, nrow(DT)))
 rm(DTn, DTl, DT)
+
+# insert columns by group in freshly unserialized data.tables, #7498
+DT = unserialize(serialize(as.data.table(mtcars), NULL))
+test(2353, DT[,foo:=mean(mpg),by=cyl], as.data.table(mtcars)[,foo:=mean(mpg),by=cyl])
```
Comment (HughParsonage) [inline inst/tests/tests.Rraw:21922]: Add a test that ensure column deletion by `NULL` as value (rather than `NULL` syntactically) is still supported

```r
DT = unserialize(serialize(as.data.table(mtcars), NULL))
null_in_value <- NULL
test(2353a, "cyl" %in% names(DT[, cyl := null_in_value]), FALSE)
DT = unserialize(serialize(as.data.table(mtcars), NULL))
test(2353b,  ncol(DT[, c("cyl", "mpg") := .(null_in_value, null_in_value)]), ncol(mtcars) - 2)
```

Comment (ben-schwen) [reply inst/tests/tests.Rraw:21922]: I added those to #7500

File: R/data.table.R [hunk 3]
```diff
@@ -1440,6 +1440,14 @@ replace_dot_alias = function(e) {
...
         lhs = names_x[m]
       }
+      # cater for deleting columns by assigning NULL
+      if ((is.null(jval) || (is.list(jval) && any(vapply_1b(jval, is.null)))) && selfrefok(x, verbose=FALSE) < 1L) {
+        name = substitute(x)
+        setalloccol(x, verbose=FALSE)
+        if (is.name(name)) {
+          assign(as.character(name), x, parent.frame(), inherits=TRUE)
+        }
+      }
```
Comment (aitap) [inline R/data.table.R:1450]: Thank you! With this commit, the 13 packages flagged for this issue pass their R CMD checks with nothing worse than NOTE.

Comment (ben-schwen) [reply R/data.table.R:1450]: Nice! I guess we need to battle test this throughly but as a bandaid it should hold
